### ISSUE TITLE
Router removes and non-standard error fields in responses

### ISSUE THREAD

**Issue by lleadbet** [apollographql/router#2110]
**Describe the bug**
Given a subgraph returning an error structure such as:

```json
{
    "errors":[
        {
            "code": "TEST_CODE",
            "message":"Code isn't a defined root field value, but okay according to the specification..."
        }
    ],
    "data":{
        "field":null
    }
}
```

The router would return:
```json
{
  "data": {
    "generateError": null
  },
  "errors": [
    {
      "message": "Code isn't a defined root field value, but okay according to the specification..."
    }
  ]
}
```

Note that it strips the root level `code` field; while this is _discouraged_ according to the spec (https://spec.graphql.org/June2018/#sec-Errors:~:text=Previous%20versions%20of%20this%20spec) it is still allowable. As such- I'd expect it to be included in the response as a result. 

**To Reproduce**
Given this sample "subgraph" in Node using express:

```js
const express = require('express')
const app = express()
const port = 4001

let resp = {
    "errors":[
        {
            "code": "TEST_CODE",
            "message":"Code isn't a defined root field value, but okay according to the specification..."
        }
    ],
    "data":{
        "field":null
    }
}

app.post('/', (req, res) => {
    res.setHeader('content-type', 'application/json')
    res.send(JSON.stringify(resp))
})

app.listen(port, () => {
  console.log(`Example app listening on port ${port}`)
})
```

And a router serving a supergraph such as: 

```graphql
schema
  @link(url: "https://specs.apollo.dev/link/v1.0")
  @link(url: "https://specs.apollo.dev/join/v0.2", for: EXECUTION)
{
  query: Query
}

directive @join__field(graph: join__Graph!, requires: join__FieldSet, provides: join__FieldSet, type: String, external: Boolean, override: String, usedOverridden: Boolean) repeatable on FIELD_DEFINITION | INPUT_FIELD_DEFINITION

directive @join__graph(name: String!, url: String!) on ENUM_VALUE

directive @join__implements(graph: join__Graph!, interface: String!) repeatable on OBJECT | INTERFACE

directive @join__type(graph: join__Graph!, key: join__FieldSet, extension: Boolean! = false, resolvable: Boolean! = true) repeatable on OBJECT | INTERFACE | UNION | ENUM | INPUT_OBJECT | SCALAR

directive @link(url: String, as: String, for: link__Purpose, import: [link__Import]) repeatable on SCHEMA

scalar join__FieldSet

enum join__Graph {
  _1_ @join__graph(name: "1", url: "http://localhost:4001")
}

scalar link__Import

enum link__Purpose {
  """
  `SECURITY` features provide metadata necessary to securely resolve fields.
  """
  SECURITY

  """
  `EXECUTION` features provide metadata necessary for operation execution.
  """
  EXECUTION
}

type Query
  @join__type(graph: _1_)
{
  q: String!
}
```

Make a request to the `q` field, and note the response. 

**Expected behavior**

The router returns: 

```json
{
  "data": {
    "generateError": null
  },
  "errors": [
    {
      "message": "Code isn't a defined root field value, but okay according to the specification...",
      "code": "TEST_CODE"
    }
  ]
}
```

**Desktop (please complete the following information):**
 - OS: MacOS
 - Version 1.4.0

**Comment by lleadbet** [comment]
Note: this is without the subgraph redaction enabled via:
```yaml
include_subgraph_errors:
  all: true # Propagate errors from all subraphs
  ```

**Comment by o0Ignition0o** [comment]
It's worth noting that we should at least add an exception Object to the error structure, to provide parity with [AS4](https://www.apollographql.com/docs/apollo-server/migration/#errorextensionsexception-is-removed)

**Comment by o0Ignition0o** [comment]
Relevant [specification](https://spec.graphql.org/October2021/#sec-Errors.Error-result-format) sections:

```
GraphQL services may provide an additional entry to errors with key extensions. This entry, if set, must have a map as its value. This entry is reserved for implementors to add additional information to errors however they see fit, and there are no additional restrictions on its contents.
```

```
GraphQL services should not provide any additional entries to the error format since they could conflict with additional entries that may be added in future versions of this specification.
```

```
 Previous versions of this spec did not describe the extensions entry for error formatting. While non-specified entries are not violations, they are still discouraged.
```

Let's see if / where we can document it.

**Comment by BrynCooke** [comment]
@o0Ignition0o @abernix The spec language of `should not` and `not violations` to me says that we should pass any extra error structures through (assuming the router is configured to pass through errors).

If the language said `must not` then would agree that we should strip.

Is there a reason why we deviate from the spec in the first place?

**Comment by abernix** [comment]
Assuming the Router is set to `include_subgraph_errors` true, I generally agree with following the guidance of the specification. I would flag that the specification doesn't make all the considerations for Federation though.  It's certainly important that we specify the behavior that should be expected, for **all** types of subgraph errors.  For example, this could get grouped into a more holistic top-down consideration along with   https://github.com/apollographql/router/issues/2422, which needs to be considerate of whether we're leaking implementation details of the federated graph or internal-transport failures to the clients.

**Comment by abernix** [comment]
Also, it would be potentially worth setting up a conversation with @IvanGoncharov about this.

**Comment by BrynCooke** [comment]
In the end any passthough of error messages is a leak of implementation details of the subgraph, so I'm not sure the argument holds that stripping extra fields is more secure.

**Comment by abernix** [comment]
I do want to emphasize that I think I agree that you're likely headed in the right direction, but —  unless I'm missing something that makes this worth considering in isolation — I believe this should be thorough top-down consideration of the experience with supporting evidence and consideration of all cases that results in a specification of behavior.  The number of considerations is bounded so, for example, making a flow chart and talking about each of them to make sure it's a holistic experience that matches the expectations of the configuration should be doable.

As for your claim above, I don't agree that any passthrough of error messages is a leak of implementation details.  A federated graph's fields might be returned by a subgraph, but that doesn't mean that the implementation details of the graph are leaked.  An error message might be soured by a particular subgraph, but the end-user wouldn't be able to derive whether they were from "subgraphA" or "subgraphB".  Some particular conditions, would be much more explicitly leaking graph-implementation details like query-planner derived field names, fetches for `@key("{ ... }")` sets, or network-transport level errors.  I don't however agree that any error message passed through from a subgraph is merely a leak though.

**Comment by abernix** [comment]
Agree, and I think there's a subgraph propagation narrative to be crafted for that, including what happens for Studio.

Some candidates to maybe consider together:

- https://github.com/apollographql/router/issues/3182
- https://github.com/apollographql/router/issues/2018
- https://github.com/apollographql/router/issues/2422

Maybe there's a themed spike to do on this that's worth prioritizing?

**Comment by o0Ignition0o** [comment]
Rereading https://github.com/apollographql/router/issues/2422#issuecomment-1697730232
I wonder if there is a compromise where we could be more permissive with errors that have a valid path. This is certainly worth considering while we're thinking of error handling in general.

**Comment by goto-bus-stop** [comment]
Not planned for the time being. This seems to be a legacy behaviour that becomes less relevant by the day; I'd strongly recommend any future reader to use the `extensions` field instead.
